File: 20-actors/sentei/methods/test_prune.py
"""test_prune.py — sentei 剪定 structural-invariant tests (stdlib unittest, ADR-2606072000).

Run: python3 -m unittest 20-actors/sentei/methods/test_prune.py
"""

from __future__ import annotations

import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from prune import (  # noqa: E402
    prune, regraft, apply_log, assert_no_verdict, to_datoms, PRUNE_ACTIONS,
)

AT = "2026-06-07T20:00:00.000Z"
COUNCIL = "did:web:etzhayyim.com:council#5of7"
MANIFESTED = {"id": "kanae.fundFlowEdge:jp-mext-2024-outlay", "manifested": True}
BUD = {"id": "kanae.fundFlowEdge:jp-mext-2025-draft", "manifested": False}


class TestG1NoPriorRestraint(unittest.TestCase):
    def test_cannot_prune_unmanifested_branch(self):
        # The core invariant: the Council cannot stop a branch before it grows.
        with self.assertRaises(ValueError):
            prune(BUD, "quarantine", "anything", COUNCIL, at=AT, council_level=7)

    def test_can_prune_after_it_manifests(self):
        p = prune(MANIFESTED, "quarantine", "G10 review", COUNCIL, at=AT, council_level=6)
        self.assertEqual(p["action"], "quarantine")
        self.assertEqual(apply_log([p]), {MANIFESTED["id"]: "quarantine"})


class TestG2AppendOnly(unittest.TestCase):
    def test_prune_appends_never_deletes(self):
        p = prune(MANIFESTED, "retract", "figure shown false", COUNCIL, at=AT, council_level=6)
        # the event is data; folding it yields state, the branch id still exists in history
        self.assertEqual(p["kind"], "prune")
        self.assertIn(MANIFESTED["id"], apply_log([p]))

    def test_as_of_time_travel_over_prune_history(self):
        p = prune(MANIFESTED, "quarantine", "g10", COUNCIL, at="2026-06-07T20:00:00.000Z", council_level=6)
        r = regraft(p, COUNCIL, at="2026-06-08T09:00:00.000Z")
        self.assertEqual(apply_log([p, r])[MANIFESTED["id"]], "live")              # current: healed
        self.assertEqual(apply_log([p, r], as_of="2026-06-07")[MANIFESTED["id"]], "quarantine")  # past: pruned
        self.assertEqual(apply_log([p, r], as_of="2026-06-05"), {})                # before: nothing yet


class TestG3GrowthUnstoppable(unittest.TestCase):
    def test_no_halt_or_delete_action(self):
        for bad in ("delete", "halt-organism", "prior-restraint", "verdict"):
            self.assertNotIn(bad, PRUNE_ACTIONS)
            with self.assertRaises(ValueError):
                prune(MANIFESTED, bad, "x", COUNCIL, at=AT, council_level=7)


class TestG4G7TransparentCare(unittest.TestCase):
    def test_verdict_token_rejected(self):
        with self.assertRaises(ValueError):
            assert_no_verdict("this branch is 違法 and the awardee is guilty")
        with self.assertRaises(ValueError):
            prune(MANIFESTED, "revoke", "punish the 犯罪", COUNCIL, at=AT, council_level=7)

    def test_basis_required(self):
        with self.assertRaises(ValueError):
            prune(MANIFESTED, "quarantine", "   ", COUNCIL, at=AT, council_level=6)

    def test_non_adjudicating_flag(self):
        p = prune(MANIFESTED, "quarantine", "g10 review", COUNCIL, at=AT, council_level=6)
        self.assertTrue(p["nonAdjudicating"])


class TestG5NoServerKey(unittest.TestCase):
    def test_server_cannot_sign_prune(self):
        with self.assertRaises(ValueError):
            prune(MANIFESTED, "quarantine", "g10", "server", at=AT, council_level=7)

    def test_prune_serverheldkey_false(self):
        p = prune(MANIFESTED, "quarantine", "g10", COUNCIL, at=AT, council_level=6)
        self.assertFalse(p["serverHeldKey"])


class TestCouncilFloorAndVote(unittest.TestCase):
    def test_invariant_adjacent_needs_lv7(self):
        with self.assertRaises(ValueError):
            prune(MANIFESTED, "rollback", "§2(a) hit", COUNCIL, at=AT, council_level=6, invariant_adjacent=True)
        ok = prune(MANIFESTED, "rollback", "§2(a) hit", COUNCIL, at=AT, council_level=7, invariant_adjacent=True)
        self.assertEqual(ok["councilLevel"], 7)

    def test_contested_prune_needs_majority(self):
        with self.assertRaises(ValueError):
            prune(MANIFESTED, "quarantine", "contested", COUNCIL, at=AT, council_level=6,
                  contested_vote={"yes": 2, "no": 5})
        ok = prune(MANIFESTED, "quarantine", "contested", COUNCIL, at=AT, council_level=6,
                   contested_vote={"yes": 6, "no": 1})
        self.assertEqual(ok["action"], "quarantine")


class TestG6Reversible(unittest.TestCase):
    def test_regraft_heals(self):
        p = prune(MANIFESTED, "quarantine", "mistaken cut", COUNCIL, at=AT, council_level=6)
        r = regraft(p, COUNCIL, at="2026-06-09T00:00:00.000Z")
        self.assertEqual(r["kind"], "regraft")
        self.assertEqual(apply_log([p, r])[MANIFESTED["id"]], "live")

    def test_server_cannot_regraft(self):
        p = prune(MANIFESTED, "quarantine", "x", COUNCIL, at=AT, council_level=6)
        with self.assertRaises(ValueError):
            regraft(p, "server", at=AT)


class TestDatomSerialization(unittest.TestCase):
    def test_to_datoms_namespace(self):
        p = prune(MANIFESTED, "quarantine", "g10", COUNCIL, at=AT, council_level=6)
        datoms = to_datoms(p)
        self.assertTrue(all(d["a"].startswith(":sentei.prune/") for d in datoms))
        self.assertTrue(any(d["a"] == ":sentei.prune/action" for d in datoms))


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: 20-actors/sentei/methods/test_prune.clj
#!/usr/bin/env bb
;; Clojure mirror of methods/test_prune.py — sentei 剪定 structural-invariant tests.
(ns sentei.methods.test-prune
  "1:1 port of the prune.py invariant suite: G1 no-prior-restraint, G2 append-only +
  as-of time-travel, G3 growth-unstoppable, G4/G7 transparent-care (no verdict token),
  G5 no-server-key, Council floor + contested vote, G6 reversible, datom serialization.
  Also pins the sha256 prune-id digest to parity with python (prune.py `_digest`).

  Run:  bb --classpath 20-actors 20-actors/sentei/methods/test_prune.clj"
  (:require [sentei.methods.prune :as p]
            [clojure.test :refer [deftest is run-tests]]))

(def AT "2026-06-07T20:00:00.000Z")
(def COUNCIL "did:web:etzhayyim.com:council#5of7")
(def MANIFESTED {"id" "kanae.fundFlowEdge:jp-mext-2024-outlay" "manifested" true})
(def BUD {"id" "kanae.fundFlowEdge:jp-mext-2025-draft" "manifested" false})

(defn- raises? [f] (try (f) false (catch Exception _ true)))

;; ── G1 no-prior-restraint ──
(deftest g1-cannot-prune-unmanifested
  (is (raises? #(p/prune BUD "quarantine" "anything" COUNCIL {:at AT :council-level 7}))))

(deftest g1-can-prune-after-manifest
  (let [d (p/prune MANIFESTED "quarantine" "G10 review" COUNCIL {:at AT :council-level 6})]
    (is (= "quarantine" (get d "action")))
    (is (= {(get MANIFESTED "id") "quarantine"} (p/apply-log [d])))))

;; ── digest parity with python ──
(deftest digest-parity
  (let [d (p/prune MANIFESTED "quarantine" "G10 review" COUNCIL {:at AT :council-level 6})]
    (is (= "prune:kanae.fundFlowEdge:jp-mext-2024-outlay:quarantine:a4067d4a7c437fcd7b50201e"
           (get d "id")))))

;; ── G2 append-only + as-of time-travel ──
(deftest g2-as-of-time-travel
  (let [d (p/prune MANIFESTED "quarantine" "g10" COUNCIL {:at "2026-06-07T20:00:00.000Z" :council-level 6})
        r (p/regraft d COUNCIL {:at "2026-06-08T09:00:00.000Z"})]
    (is (= "live"       (get (p/apply-log [d r]) (get MANIFESTED "id"))))            ; current: healed
    (is (= "quarantine" (get (p/apply-log [d r] "2026-06-07") (get MANIFESTED "id")))) ; past: pruned
    (is (= {}           (p/apply-log [d r] "2026-06-05")))))                          ; before: nothing

;; ── G3 growth-unstoppable ──
(deftest g3-no-halt-or-delete-action
  (doseq [bad ["delete" "halt-organism" "prior-restraint" "verdict"]]
    (is (not (contains? p/PRUNE-ACTIONS bad)))
    (is (raises? #(p/prune MANIFESTED bad "x" COUNCIL {:at AT :council-level 7})))))

;; ── G4/G7 transparent care ──
(deftest g7-verdict-token-rejected
  (is (raises? #(p/assert-no-verdict "this branch is 違法 and the awardee is guilty")))
  (is (raises? #(p/prune MANIFESTED "revoke" "punish the 犯罪" COUNCIL {:at AT :council-level 7}))))

(deftest g7-basis-required
  (is (raises? #(p/prune MANIFESTED "quarantine" "   " COUNCIL {:at AT :council-level 6}))))

(deftest g7-non-adjudicating-flag
  (is (true? (get (p/prune MANIFESTED "quarantine" "g10 review" COUNCIL {:at AT :council-level 6})
                  "nonAdjudicating"))))

;; ── G5 no-server-key ──
(deftest g5-server-cannot-sign
  (is (raises? #(p/prune MANIFESTED "quarantine" "g10" "server" {:at AT :council-level 7})))
  (is (false? (get (p/prune MANIFESTED "quarantine" "g10" COUNCIL {:at AT :council-level 6})
                   "serverHeldKey"))))

;; ── Council floor + contested vote ──
(deftest council-invariant-needs-lv7
  (is (raises? #(p/prune MANIFESTED "rollback" "§2(a) hit" COUNCIL
                         {:at AT :council-level 6 :invariant-adjacent true})))
  (is (= 7 (get (p/prune MANIFESTED "rollback" "§2(a) hit" COUNCIL
                         {:at AT :council-level 7 :invariant-adjacent true})
                "councilLevel"))))

(deftest contested-prune-needs-majority
  (is (raises? #(p/prune MANIFESTED "quarantine" "contested" COUNCIL
                         {:at AT :council-level 6 :contested-vote {"yes" 2 "no" 5}})))
  (is (= "quarantine" (get (p/prune MANIFESTED "quarantine" "contested" COUNCIL
                                    {:at AT :council-level 6 :contested-vote {"yes" 6 "no" 1}})
                           "action"))))

;; ── G6 reversible ──
(deftest g6-regraft-heals
  (let [d (p/prune MANIFESTED "quarantine" "mistaken cut" COUNCIL {:at AT :council-level 6})
        r (p/regraft d COUNCIL {:at "2026-06-09T00:00:00.000Z"})]
    (is (= "regraft" (get r "kind")))
    (is (= "live" (get (p/apply-log [d r]) (get MANIFESTED "id")))))
  (let [d (p/prune MANIFESTED "quarantine" "x" COUNCIL {:at AT :council-level 6})]
    (is (raises? #(p/regraft d "server" {:at AT})))))

;; ── datom serialization ──
(deftest to-datoms-namespace
  (let [d (p/prune MANIFESTED "quarantine" "g10" COUNCIL {:at AT :council-level 6})
        datoms (p/to-datoms d)]
    (is (every? #(clojure.string/starts-with? (get % "a") ":sentei.prune/") datoms))
    (is (some #(= (get % "a") ":sentei.prune/action") datoms))
    (is (some #(= (get % "v_edn") "\"quarantine\"") datoms))))

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'sentei.methods.test-prune)]
    (System/exit (if (zero? (+ fail error)) 0 1))))
